File: Python_Code/MyShop.py
# Method to show the Product List
def get_productList(product):
    print("product name : Price")
    print()
    for key, value in product.items():
        print(key, ":", value)
    print()


# Here we are colleting the oder details from the user
def get_product_Details(num_product, product):
    count = int(input("please Enter How many product you want to buy : ").strip())
    cartList = []
    if count <= num_product:
        for i in range(count):
            PName = input("Enter the product Name you want to buy : ").strip()
            if PName not in product:
                print("please enter valid product Name")
                print()
                PName = input("Enter the product Name you want to buy : ").strip()
            print()
            Pquant = input("Please inter the number of unites you want to buy : ").strip()

            if int(Pquant) <= 0:
                print("plese enter the valid number of unit")
                print()
                Pquant = input("Please inter the number of unites you want to buy : ").strip()
            print()
            isgiftwrap = int(
                input(
                    "Do we have to wrap it as gift please enter 1 for yes and 0 for No : "
                ).strip()
            )
            print()
            cartList.append([PName, int(Pquant), bool(isgiftwrap)])
    
        return cartList
    else:
        print("Please enter valid number of product")
        return 0 


# Here we are calculating processing the oder details
def calculateDetails(detailList, product):
    detail_dict = {
        "total_cost": 0,
        "total_quantity": 0,
        "individual_Product_quantity": {},
        "individual_Product_bool": {},
        "individual_total_price": {},
    }

    for p_details in detailList:
        Pname, Pquantity, isgiftwrap = p_details
        unit_price = product.get(Pname)
        total_price = unit_price * Pquantity

        detail_dict["individual_total_price"][Pname] = total_price
        detail_dict["total_cost"] += total_price
        detail_dict["total_quantity"] += Pquantity
        detail_dict["individual_Product_quantity"][Pname] = Pquantity
        detail_dict["individual_Product_bool"][Pname] = isgiftwrap

    return detail_dict


# here we are calculating the maximum discount on the product
def discountCalculator(details, Original_product):
    total_cost = details["total_cost"]
    total_quantity = details["total_quantity"]
    individu_quanity = details["individual_Product_quantity"]
    individual_value = details["individual_Product_bool"]
    individual_total_price = details["individual_total_price"]

    discount_rule = {
        "flat_10_discount": total_cost > 200,
        "bulk_5_discount": any(quantity > 10 for quantity in individu_quanity.values()),
        "bulk_10_discount": total_quantity > 20,
        "tiered_50_discount": total_quantity > 30
        and any(quantity > 15 for quantity in individu_quanity.values()),
    }
    flat_10_discount_list = []
    bulk_5_discount_list = []
    bulk_10_discount_list = []
    tiered_50_discount_list = []

    applicable_discounts = [
        rule for rule, apply_condition in discount_rule.items() if apply_condition
    ]

    if not applicable_discounts:
        return 0

    for rule in applicable_discounts:
        if rule == "flat_10_discount":
            flat_10_discount_list.append(10)
            # discount_amount.append({"flat_10_discount": 10})
        elif rule == "bulk_5_discount":
            for product, quantity in individu_quanity.items():
                if quantity > 10:
                    bulk_5_discount_list.append(0.05 * individual_total_price[product])
                    # discount_amount.append({"bulk_5_discount": 0.05 * individual_total_price[product]})
        elif rule == "bulk_10_discount":
            bulk_10_discount_list.append(0.1 * total_cost)
            # discount_amount.append({"bulk_10_discount":0.1 * total_cost})
        elif rule == "tiered_50_discount":
            for product, quantity in individu_quanity.items():
                if quantity > 15:
                    above_15_amount = (quantity - 15) * Original_product[product]
                    tiered_50_discount_list.append(0.5 * above_15_amount)
                    # discount_amount.append({"tiered_50_discount":0.5 * individual_total_price[product]})
    discount_amount= {}
    # some discount list may be empty also if we are selecting one product 
    if flat_10_discount_list :
        discount_amount["flat_10_discount"]  = max(flat_10_discount_list)
        
    if bulk_5_discount_list:
        discount_amount["bulk_5_discount"] = max(bulk_5_discount_list)
    if bulk_10_discount_list:
        discount_amount["bulk_10_discount"] = max(bulk_10_discount_list)
    if tiered_50_discount_list :
        discount_amount["tiered_50_discount"] = max(tiered_50_discount_list) 

    # discount_amount = {
    #     "flat_10_discount": max(flat_10_discount_list),
    #     "bulk_5_discount": max(bulk_5_discount_list),
    #     "bulk_10_discount": max(bulk_10_discount_list),
    #     "tiered_50_discount": max(tiered_50_discount_list),
    # }
    max_discount = 0
    applicable_rule = ""
    for dicount_rule, amount in discount_amount.items():
        if amount > max_discount:
            max_discount = amount
            applicable_rule = dicount_rule
    return [max_discount, applicable_rule]


# here we are shipping and figtwrap fee
def calculate_page_and_giftrap_charge(total_quantity, isgiftwrap, individual_Product_quantity):
    gift_wrap_fee = 1
    Number_of_unit_inPackage = 10
    shipping_fee = 5
    sum = 0
    for product, unit in isgiftwrap.items():
        if unit:
            sum += individual_Product_quantity[product]

    total_gift_wrap_fee = sum * gift_wrap_fee
    if total_quantity <= 10:
        total_shipping_fee = 10
    else:
        total_shipping_fee = (total_quantity / Number_of_unit_inPackage) * shipping_fee
    return [total_gift_wrap_fee, total_shipping_fee]


#  here we are printing the Oder Summary
def ShowSummary(details, discount, fee):
    individual_Product_quantity = details["individual_Product_quantity"]
    individual_total_price = details["individual_total_price"]
    sub_total = 0
    print("******* Your Oder Summery *********")
    print()
    for product, value in individual_Product_quantity.items():
        print("Product Name", ":", product)
        print("Units Bought: ", value)
        print("total amount of the product:", individual_total_price[product])
        print()
        sub_total += individual_total_price[product]

    print("Sub TOtal: ", sub_total)
    print()

    # validating the discount 
    if discount == 0:
        print("please buy some more products to get Discount")
    else:
        print("Applied Discount : ", discount[1], ", total discount : ", discount[0])
    print()
    print("gift wraping fee: ", fee[0])
    print()
    print("Shipping Charge: ", fee[1])
    print()
    if discount == 0:
        print("Total : ", sub_total)
    else:
        print("Total : ", sub_total - discount[0])
    print()


if __name__ == "__main__":
    print()
    
    print("+++++++++ Welcome To Our Shop ++++++++++")

    product = {"ProductA": 20, "ProductB": 40, "ProductC": 50}
    num_product = len(product)
    get_productList(product)
    detailList = get_product_Details(num_product, product)

    details = calculateDetails(detailList, product)

    max_discount = discountCalculator(details, product)

    fee = calculate_page_and_giftrap_charge(
        details["total_quantity"],
        details["individual_Product_bool"],
        details["individual_Product_quantity"],
    )

    ShowSummary(details, max_discount, fee)

    print("+++++++++ Thank You For Shoping Please Visit Again ++++++++++")
    print()

File: Python_Code/MyShop.js

// Method to show the Product List
function get_productList(product) {
    console.log("product name : Price");
    for (const Pname in product){
        console.log(Pname, ":" ,product[Pname])
    }
}

// Collecting order details from the user
function get_product_details(Number_of_product, product) {
    const count = parseInt(prompt("Please Enter How many products you want to buy : ").trim());
    const cartList = [];
    console.log(Number_of_product, count,"NUm")
    if (count <= Number_of_product){

        for (let i = 0; i < count; i++) {
            let PName = prompt("Enter the product Name you want to buy : ").trim();

            // cheking that enterd product is present or not 
            if( ! product.hasOwnProperty(PName)){
                  alert("Please Enter a vailid product")
                  PName = prompt("Enter the product Name you want to buy : ").trim();
 
            }
            let Pquant = parseInt(prompt("Please enter the number of units you want to buy : ").trim());

            // cheking that number or units is greater then 0 or not
            if(Pquant <= 0){
                alert("Please enter a valid Units ")
                Pquant = parseInt(prompt("Please enter the number of units you want to buy : ").trim());
            }
            const isgiftwrap = parseInt(prompt("Do we have to wrap it as a gift? Please enter 1 for yes and 0 for No : ").trim());

            cartList.push([PName, Pquant, Boolean(isgiftwrap)]);
        }
    }else{
        alert("Please enter the vailid number of product")
    }
    return cartList;
}

// Here we are proccesing the order details entered odered by user 
function calculateDetails(detailList, product) {

    // Created dictionry to structure the data 
    const detail_dict = {
        total_cost: 0,
        total_quantity: 0,
        individual_Product_quantity: {},
        individual_Product_bool: {},
        individual_total_price: {},
    };

    for (const Pdetails of detailList) {
        const [Pname, Pquantity, isgiftwrap] = Pdetails;
        const unitPrice = product[Pname];
        const totalPrice = unitPrice * Pquantity;

        detail_dict.individual_total_price[Pname] = totalPrice;
        detail_dict.total_cost += totalPrice;
        detail_dict.total_quantity += Pquantity;
        detail_dict.individual_Product_quantity[Pname] = Pquantity;
        detail_dict.individual_Product_bool[Pname] = isgiftwrap;
    }

    return detail_dict;
}

// Here we are calculating the total discount we will get ont he product
function discountCalculator(details, originalProduct) {
    const total_cost = details.total_cost;
    const total_quantity = details.total_quantity;
    const individul_quanity = details.individual_Product_quantity;
    const individual_value = details.individual_Product_bool;
    const individual_total_price = details.individual_total_price;
     

    const discountRule = {
        flat_10_discount: total_cost > 200,
        bulk_5_discount: Object.values(individul_quanity).some(quantity => quantity > 10),
        bulk_10_discount: total_quantity > 20,
        tiered_50_discount: total_quantity > 30 && Object.values(individul_quanity).some(quantity => quantity > 15),
    };

    // created list to store the same type of discount of each product 
    const flat_10_discountList = [];
    const bulk_5_discountList = [];
    const bulk_10_discountList = [];
    const tiered_50_discountList = [];
    
    // here we creating the list of all applicable discount 
    const applicableDiscounts = Object.keys(discountRule).filter(rule => discountRule[rule]);
    if (!applicableDiscounts.length) {
        console.log("not applocable")
        return 0;
    }
   // here we are calculating applicable discount in each product and storing it in list 
    for (const rule of applicableDiscounts) {
        switch (rule) {
            case 'flat_10_discount':
                flat_10_discountList.push(10);
                break;
            case 'bulk_5_discount':
                for (const [product, quantity] of Object.entries(individul_quanity)) {
                    if (quantity > 10) {
                        bulk_5_discountList.push(0.05 * individual_total_price[product]);
                    }
                }
                break;
            case 'bulk_10_discount':
                bulk_10_discountList.push(0.1 * total_cost);
                break;
            case 'tiered_50_discount':
                for (const [product, quantity] of Object.entries(individul_quanity)) {
                    if (quantity > 15) {
                        const above15Amount = (quantity - 15) * originalProduct[product];
                        tiered_50_discountList.push(0.5 * above15Amount);
                    }
                }
                break;
        }
    }

    
    //  here we are calculating the maxdiscount in each type of discount from all product 
    const discountAmount = {
        flat_10_discount: Math.max(...flat_10_discountList),
        bulk_5_discount: Math.max(...bulk_5_discountList),
        bulk_10_discount: Math.max(...bulk_10_discountList),
        tiered_50_discount: Math.max(...tiered_50_discountList),
    }; 

    // here we are calculating the max discount from all the discount
    let max_discount = 0;
    let applicableRule = '';
    for (const [discountRule, amount] of Object.entries(discountAmount)) {
        if (amount > max_discount) {
            max_discount = amount;
            applicableRule = discountRule;
        }
    }
    return [max_discount, applicableRule];
}

// Calculate shipping and gift wrap fee
function calculate_page_and_giftrap_charge(total_quantity, isgiftwrap, individual_product_quantity) {
    const gift_wrap_fee = 1;
    const number_Of_Unit_InPackage = 10;
    const shipping_Fee = 5;
    let sum = 0;
    for (const [product, unit] of Object.entries(isgiftwrap)) {
        if (unit) {
            sum += individual_product_quantity[product];
        }
    }

    const total_Gift_Wrap_Fee = sum * gift_wrap_fee;
    const total_Shipping_Fee = total_quantity <= 10 ? 10 : (total_quantity / number_Of_Unit_InPackage) * shipping_Fee;
    return [total_Gift_Wrap_Fee, total_Shipping_Fee];
}

// Print the Order Summary
function ShowSummary(details, discount, fee) {
    const individual_product_quantity = details.individual_Product_quantity;
    const individual_total_price = details.individual_total_price;
    let subTotal = 0;
    console.log("******* Your Order Summary *********");
    for (const [product, value] of Object.entries(individual_product_quantity)) {
        console.log("Product Name", ":", product);
        console.log("Units Bought: ", value);
        console.log("Total amount of the product:", individual_total_price[product]);
        subTotal += individual_total_price[product];
    }

    console.log("Sub Total: ", subTotal);

    // validating the discount 

    if (discount == 0){
        alert("please Buy some more product to get the discount")
    }else{
        console.log("Applied Discount : ", discount[1], ", total discount : ", discount[0]);
    }
    console.log("Gift wrapping fee: ", fee[0]);
    console.log("Shipping Charge: ", fee[1]);
    console.log("Total : ", subTotal - discount[0]);
}

// ############# function calling starts from here  ########## 
console.log("+++++++++ Welcome to our Shop ++++++++++");

// product list 
const product = { ProductA: 20, ProductB: 40, ProductC: 50 };

// number of product in the shop 
const Number_of_product = Object.keys(product).length; 

get_productList(product);
const detailList = get_product_details(Number_of_product, product);
const details = calculateDetails(detailList, product);
const max_discount = discountCalculator(details, product);
const fee = calculate_page_and_giftrap_charge(
    details.total_quantity,
    details.individual_Product_bool,
    details.individual_Product_quantity
);
ShowSummary(details, max_discount, fee);
console.log("+++++++++ Thank you for Shopping. Please visit again! ++++++++++");

